src/vm/types.rs

```diff
@@ -70,143 +212,399 @@ impl TupleTypeSignature {
         Ok(TupleTypeSignature { type_map: type_map })
     }
 
-    pub fn check_valid(&self, name: &str, value: &Value) -> bool {
-        if let Some(expected_type) = self.type_map.get(name) {
-            *expected_type == TypeSignature::type_of(value)
+    pub fn admits(&self, other: &TupleTypeSignature) -> bool {
+        if self.type_map.len() != other.type_map.len() {
+            return false
+        }
+
+        for (name, my_type_sig) in self.type_map.iter() {
+            if let Some(other_type_sig) = other.type_map.get(name) {
+                if !my_type_sig.admits_type(other_type_sig) {
+                    return false
+                }
+            } else {
+                return false
+            }
+        }
+
+        return true
+    }
+
+    pub fn size(&self) -> i128 {
+        let mut name_size: i128 = 0;
+        let mut value_size: i128 = 0;
+        for (name, type_signature) in self.type_map.iter() {
+            // we only accept ascii names, so 1 char = 1 byte.
+            name_size = name_size.checked_add(name.len() as i128).unwrap();
```

Comment:
For these `unwrap()`s, can you add an error message?

Reply:
Also, how certain are we that these `unwrap()`s never occur?

Reply:
Sure -- will add an error message.

These panics should never occur, since the composite values of the tuple should be smaller than MAX_VALUE_SIZE (much smaller than i128::max bytes) and combining them until the tuple is too large would require an extraordinarily large program. However, since the max value size check for tuple (and list) types occurs using the `size()` method, it's theoretically possible. So I'll change these to raise a ValueTooLarge error.